### PR TITLE
Fix wrong folding behaviour of last block if containing comment

### LINKED ISSUE #52326: Comment causes contents of folded methods to not hide (CodeEdit regression)
### Godot version

v4.0.dev.calinou.036b7a098

### System information

Windows 10, Vulkan, Nvidia GeForce RTX 2060 Max-Q

### Issue description

When adding a comment in a method, any lines of code after the comment are not hidden when the method is folded. For instance, if you write a line of code, then write a comment on the next line to describe what the code does, then write some more code below the comment, and code below the comment will not be hidden when the method is folded. Images below:

image
image

### Steps to reproduce

1. Make a new script in Godot 4.0
2. In the `_ready()` function (or any other function), write a line of code
3. Below that code, make a comment
4. Below the comment, write another line of code
5. Fold the `_ready()` function

**Note:** I've tried this on arrays. It doesn't seem to be an issue with them.

### Minimal reproduction project

[Method Folding Issue.zip](https://github.com/godotengine/godot/files/7094375/Method.Folding.Issue.zip)

### LINKED ISSUE #52385: Out of bounds error when folding last line of code
### Godot version

v4.0.dev.custom_build [ade4e9320]

### System information

Windows 10

### Issue description

Folding a block of code at the end of a script file cause an out of bounds error in the console.

### Steps to reproduce

1 . Create a script : 

image

2 . Fold the block containing the last line of script file.

image

3. Error appears in console (would be p_line 4 with the script above)

out_of_bound_error

### Minimal reproduction project

[FOLD CODE.zip](https://github.com/godotengine/godot/files/7108637/FOLD.CODE.zip)

### PR DESCRIPTION
**Pull request by jmb462**
Fix #52326
Fix #52385

**Note : This first description is now obsolete. See comment here to new implementation : https://github.com/godotengine/godot/pull/52339#issuecomment-913032405**

### Issue description

When adding a comment in a method, any lines of code after the comment are not hidden when the method is folded.

### Fix proposal

As I can't find a case where it won't be OK, I suggest to ignore comment at same or higher indentation level as a cause of end folding condition.

### Before

fold_bug

### After

indented

### CODE REVIEW

**Review by Paulb23 — CHANGES_REQUESTED** [review]
This check is to handle the second case in the issue where the comment or string does not follow the indent, but despite that should still be folded.

**Comment by jmb462** [comment]
@Paulb23 Do you mean this case ? 

unindented

Neither the current code in master nor my PR fold after the comment in that case.
I could modify to deal with it.

Or do you mean another case ?

**Comment by Paulb23** [comment]
> Do you mean this case ?

Yes.

**Comment by name-here** [comment]
I think expected behavior would be to end folding on a comment *only* if the comment is at the end of the indented block (all *code* after it is less indented), and the comment is less indented than the block you're folding.  Someone doing their commenting in a funky way (say, `#[tab][tab] [comment]` instead of `[tab][tab] #[comment]` shouldn't reasonably mess up folding.  The exception for less-indented comments at the end of folding code blocks won't *always* match user intention, but I think it's the best we can do without reading the user's mind.

**Comment by jmb462** [comment]
@name-here Could you post a short script example that illustrate the case and what would be the expected behaviour.

I'm comparing with folding in visual studio / vscode or other code editor to try to find a "standard" behavior.

**Comment by jmb462** [comment]
The bug wasn't correctly identified in the issue.
It was not just a problem with comments but more with end of script : The end of script doesn't fold if there is a comment in this folded block.

So first fix proposed here is obsolete.

I force pushed the new way to fix the issue.

### Identified cause
The parse of the script never catches the case were the folded block contains a comment and is the last block of script. (end_line variable remain at last comment index instead of last line of script).

### Fix proposed
Add a test to update end_line to the last line of script in that case.

Also, I've compared with the folding behaviour of VS, VSCode and SublimeText.
All these editors don't folder the last empty line of block to keep a space between the folded block and the following block.
Seems to be standard behavior and more readable I think.

### Before

You can see that the first and second function are the same. But the one at the end of script doesn't fold correctly.
And no space is keep between folded function and the following one.
buged

### After

The two blocks folds correctly. An empty line is kept unfolded after folded block.
fixed

File: scene/gui/code_edit.cpp [hunk 1]
```diff
@@ -1400,15 +1400,16 @@ void CodeEdit::fold_line(int p_line) {
 	}
 
 	/* Find the last line to be hidden. */
-	int end_line = get_line_count();
+	int last_line = get_line_count() - 1;
```
Comment (Paulb23) [inline scene/gui/code_edit.cpp:1403]: Would suggest renaming this to `line_count` or  `text_size` to be little clearer. Could also make it `const`.

File: scene/gui/code_edit.cpp [hunk 2]
```diff
@@ -1417,19 +1418,33 @@ void CodeEdit::fold_line(int p_line) {
...
+					/* Keep an empty line unfolded if any */
+					end_line = i - 2;
+				} else {
+					end_line = i - 1;
+				}
 				break;
 			}
 		}
 	}
-
```
Comment (Paulb23) [inline scene/gui/code_edit.cpp:1432]: Would keep this blank line.

File: scene/gui/code_edit.cpp [hunk 3]
```diff
@@ -1417,19 +1418,33 @@ void CodeEdit::fold_line(int p_line) {
...
 
-			if (get_indent_level(i) <= start_indent && get_line(i).strip_edges().size() != 0) {
-				end_line = i - 1;
+			if (i == last_line) {
+				/* Do not fold empty last line of script if any */
+				if (get_line(i).strip_edges().size() != 0) {
+					end_line = i;
+				} else {
+					end_line = i - 1;
+				}
```
Comment (Paulb23) [inline scene/gui/code_edit.cpp:1433]: Nitpick: 
```suggestion
				end_line = i;
				if (get_line(i).strip_edges().size() == 0) {
					end_line--;
				}
```

File: scene/gui/code_edit.cpp [hunk 4]
```diff
@@ -1417,19 +1418,33 @@ void CodeEdit::fold_line(int p_line) {
...
+				break;
+			}
+
+			if ((get_indent_level(i) <= start_indent && get_line(i).strip_edges().size() != 0)) {
+				if (get_line(i - 1).strip_edges().size() == 0 && i - 2 > p_line) {
+					/* Keep an empty line unfolded if any */
+					end_line = i - 2;
+				} else {
+					end_line = i - 1;
+				}
```
Comment (Paulb23) [inline scene/gui/code_edit.cpp:1443]: Nitpick:
```suggestion
				/* Keep an empty line unfolded if any */
				end_line = i - 1;
				if (get_line(i - 1).strip_edges().size() == 0 && i - 2 > p_line) {
					end_line--;
				}
```

**Review by Paulb23 — CHANGES_REQUESTED** [review]
Looks good, couple small changes to make it a little clearer.

As an aside, might be worth closing #52386 and just having this one as it should fix both issues.

File: scene/gui/code_edit.cpp (hunk 2 above)
Comment (jmb462) [reply scene/gui/code_edit.cpp]: This change was unintentionnal. Sorry.

**Comment by jmb462** [comment]
Thanks for the help.
Proposed changes has been applied.

I'm closing other PR as you suggested.

**Review by Paulb23 — APPROVED** [review]

**Comment by jmb462** [comment]
I've just added ``const`` to last_line variable declaration after your approval.

**Review by mhilbrunner — APPROVED** [review]

**Comment by mhilbrunner** [comment]
Thanks!

**Pull request merged**